IO/d4est_norms.c

```diff
@@ -0,0 +1,617 @@
+  if (energy_norm_data != NULL)
+    local_energy_norm_sqr = d4est_ip_energy_norm_compute
+                            (
+                             p4est,
+                             error,
+                             energy_norm_data,
+                             ghost,
+                             ghost_data,
+                             d4est_ops,
+                             d4est_geom,
+                             d4est_quad,
+                             d4est_factors
+                            );
+  
+  double local_nodes_dbl = (double)local_nodes;
+  double local_quad_nodes_dbl = (double)local_quad_nodes;
+  double local_estimator = estimator;
+  double local_reduce [5];
+  double global_reduce [5];
+  double global_Linf;
+    
+  local_reduce[0] = local_l2_norm_sqr;
+  local_reduce[1] = local_nodes_dbl;
+  local_reduce[2] = local_quad_nodes_dbl;
+  local_reduce[3] = local_estimator;
+  local_reduce[4] = local_energy_norm_sqr;
+    
+  sc_reduce
+    (
+     &local_reduce[0],
+     &global_reduce[0],
+     5,
+     sc_MPI_DOUBLE,
+     sc_MPI_SUM,
+     0,
+     sc_MPI_COMM_WORLD
+    );
+
+  sc_reduce
+    (
+     &local_Linf,
+     &global_Linf,
+     1,
+     sc_MPI_DOUBLE,
+     sc_MPI_MAX,
+     0,
+     sc_MPI_COMM_WORLD
+    );
+
+  double global_l2_norm_sqr = global_reduce[0];
+  double global_nodes_dbl = global_reduce[1];
+  double global_quad_nodes_dbl = global_reduce[2];
+  double global_estimator = global_reduce[3];
+  double global_energy_norm_sqr = global_reduce[4];
+
+  int avg_deg = pow((int)(global_nodes_dbl/p4est->global_num_quadrants), 1./(P4EST_DIM)) - 1.f + .5f;
+  int avg_deg_quad = pow((int)(global_quad_nodes_dbl/p4est->global_num_quadrants), 1./(P4EST_DIM)) - 1.f + .5f;
+
+  if(energy_norm_data != NULL && fit != NULL){
+    d4est_norms_energy_norm_add_entry_and_fit(p4est,fit, global_energy_norm_sqr, global_nodes_dbl);
+  }
+    
+  if (p4est->mpirank == 0){
+    zlog_category_t *c_norms = zlog_get_category("norms");
+    zlog_info(c_norms, "%d %d %d %d %d %.25f %.25f %.25f %.25f",
+      (int)p4est->global_num_quadrants,
+      (int)global_nodes_dbl,
+      avg_deg,
+      (int)global_quad_nodes_dbl,
+      avg_deg_quad,
+      (global_estimator < 0) ? -1. : sqrt(global_estimator),
+      (global_l2_norm_sqr < 0) ? -1 : sqrt(global_l2_norm_sqr),
+      (global_energy_norm_sqr < 0) ? -1. : sqrt(global_energy_norm_sqr),
+      global_Linf
+    );
+  }
+
+
+  return (d4est_norms_norms_t)
+    {.global_nodes = (int)global_nodes_dbl,
+        .global_num_quadrants = p4est->global_num_quadrants,
+        .avg_deg = avg_deg,
+        .global_quad_nodes = (int)global_quad_nodes_dbl,
+        .avg_deg_quad = avg_deg_quad,
+        .global_estimator = (global_estimator < 0) ? -1. : sqrt(global_estimator),
+        .global_l2_norm_sqr = (global_l2_norm_sqr < 0) ? -1 : sqrt(global_l2_norm_sqr),
+        .global_linf = global_Linf
+        };
+    
+}
+
+double
+d4est_norms_L2
+(
+  double *field_value_errors,
```

Comment:
I chose to directly hand over the pre-computed errors to the norm functions, instead of the field values plus their compare values, since a norm function usually needs to compute those anyway (all 3 implemented norm functions only depend on the error). If for some reason a custom norm function needs the original field values, they can be provided through the context.

Reply:
Smart thinking, this is way better.